### DISCUSSION TITLE
Using pre-commit for space-ros repositories

### DISCUSSION THREAD

**Discussion by Ryanf55** [space-ros/space-ros#142] [Ideas]
# pre-commit for SpaceROS

## Purpose

I would like to discuss SpaceROS using pre-commit for linting.

##  Related

See this for context: https://github.com/space-ros/space-ros/issues/80#issuecomment-1936218616

## Key Decisions

1. Do we need to share the pre-commit configuration across repos, or is it ok to have a `.pre-commit` file in each repo
1.  Which hooks do we enable
1. What is the upgrade process for hooks (each one is tied to a version)
1. Is there a reason to not run them in CI
1. Are we opposed to having pre-commit CI autofix contributions? It would add fixup commits, which would need a `squash` strategy set up for all repos
1. Are we have a mass-reformat for some of the more picky hooks (such as enforcing `\n` as a newline on all text files

## Details on decisions with my recommendations

1. See the linked comment- I propose not sharing the config across repos because it's hard to implement, and not much time to hard-code a config in each repo and it requires little maintenance
1. We should enable as many as possible, one for each type of file (IE, if we use docker, add a docker linter). Having too many hooks slows down commit speed, but even with 30 hooks, it's only about 5 seconds unless you are doing a nasty rebase. You can always ignore hooks. Perhaps start with the set in ArduPilot.
1. I propose regular upgrades of patch versions on hooks, minor version updates at each release, and no major version update until the next LTS distribution, unless reviewed by the maintainers and agreed upon by users. Major version changes are quite horrible at causing massive code divergence and making developers frustrated if they have long-running feature branches
1. I propose running in CI, and recomending developers also run the hooks locally during dev
1. I think it's good if a squash strategy is used, especially if we get lots of contributions from random people that don't want to iterate on fixes
1. We can use [git-blame-ignore-revs](https://github.blog/changelog/2022-03-24-ignore-commits-in-the-blame-view-beta/) to avoid code noise, and do a mass format, in line with approach that [GDAL took](https://gdal.org/development/rfc/rfc69_cplusplus_formatting.html). We should select a date to do this operation, and also provide instructions for people not git-experts on doing a formatting rebase with conflict resolution hints to lower the friction of the format to the bare minimum

## Timeline

We want to get this stood up by the next release. The most difficult thing is making the above decisions, the rest is quite straightforward.

**Comment by Bckempa** [comment]
Thanks for the detailed thoughts on this, @Ryanf55!

When this was first suggested I was hesitant partially because I was mentally conflating the build and test of the Space ROS owned packages with the build and test of the whole Space ROS distribution. You are correct for the individual packages we absolutely should be applying the same best-practices we'd like to see upstream and I agree using pre-commit absolutely qualifies.

> 1. Do we need to share the pre-commit configuration across repos, or is it ok to have a `.pre-commit` file in each repo
> > See the linked comment- I propose not sharing the config across repos because it's hard to implement, and not much time to hard-code a config in each repo and it requires little maintenance

Given the above, I agree that having this per-repo makes the most sense.

> 2. Which hooks do we enable
> > We should enable as many as possible, one for each type of file (IE, if we use docker, add a docker linter). Having too many hooks slows down commit speed, but even with 30 hooks, it's only about 5 seconds unless you are doing a nasty rebase. You can always ignore hooks. Perhaps start with the set in ArduPilot.

Here I'm less concerned about number or speed and more about level of detail. I have been guilty of (ab)using pre-commit as an all-in-one test coordinator but since these are all ROS packages I think that higher level stuff should be handled by ament/colcon and pre-commit should stick to basic lints and style rules.

> 3. What is the upgrade process for hooks (each one is tied to a version)
> > I propose regular upgrades of patch versions on hooks, minor version updates at each release, and no major version update until the next LTS distribution, unless reviewed by the maintainers and agreed upon by users. Major version changes are quite horrible at causing massive code divergence and making developers frustrated if they have long-running feature branches

This sounds reasonable to me. To keep the maintenance burden manageable we should require the procedure and schedule for these updates be codified when we add pre-commit and maybe make an issue template or something to simplify the process as well. 

> 4. Is there a reason to not run them in CI
> > I propose running in CI, and recomending developers also run the hooks locally during dev

I agree we should run in CI with the clarification that this should be part of the per-repo CI acceptance testing, not the build of the Space ROS distribution workspace and that we enable, recommend, and encourage, but not require, running locally.

> 5. Are we opposed to having pre-commit CI autofix contributions? It would add fixup commits, which would need a `squash` strategy set up for all repos
> > I think it's good if a squash strategy is used, especially if we get lots of contributions from random people that don't want to iterate on fixes

I do not think we should let CI make code changes, especially since this project has an unusual blend of requiring very deliberate, traceable changes from a maintainer base with limited bandwidth. Automatic fixes would mess with the existing squash-and-merge procedure as well as make it easier to have the (hopefully benign) unreviewed changes get merged accidentally.

To start I'd vote to have pre-commit added as a non-mandatory CI task so the results are visible to maintainers but not required while we tune the rule set. From there we can discuss if it makes sense to require a passing lint.

> 6. Are we have a mass-reformat for some of the more picky hooks (such as enforcing `\n` as a newline on all text files
> > We can use [git-blame-ignore-revs](https://github.blog/changelog/2022-03-24-ignore-commits-in-the-blame-view-beta/) to avoid code noise, and do a mass format, in line with approach that [GDAL took](https://gdal.org/development/rfc/rfc69_cplusplus_formatting.html). We should select a date to do this operation, and also provide instructions for people not git-experts on doing a formatting rebase with conflict resolution hints to lower the friction of the format to the bare minimum

Since this would only impact the relatively few, relatively small repos owned by the Space ROS org and are worked on by a relatively small team I don't see a reason not to have a mass-fix commit and I think we'd want to preserve the traceability of when and why those changes were made rather than hiding it. The only complication would be if the checks turn up anything we don't want to change for whatever reason, but we'll need to handle that on a case-by-case basis either way.

**Comment by Ryanf55** [comment]
> Thanks for the detailed thoughts on this, @Ryanf55!

Happy to get the ball rolling.

> When this was first suggested I was hesitant partially because I was mentally conflating the build and test of the Space ROS owned packages with the build and test of the whole Space ROS distribution. You are correct for the individual packages we absolutely should be applying the same best-practices we'd like to see upstream and I agree using pre-commit absolutely qualifies.
Yep, I see pre-commit only being checked in the repo's .github/workflows. The build farm can run colcon and bloom as usual.


> Here I'm less concerned about number or speed and more about level of detail. I have been guilty of (ab)using pre-commit as an all-in-one test coordinator but since these are all ROS packages I think that higher level stuff should be handled by ament/colcon and pre-commit should stick to basic lints and style rules.

Agree, pre-commit should be quick, and not related to the correctness of the code. Unit testing, integration testing, and the standard ROS linters can run through colcon as usual. Developers can still run these locally if they desire.


> This sounds reasonable to me. To keep the maintenance burden manageable we should require the procedure and schedule for these updates be codified when we add pre-commit and maybe make an issue template or something to simplify the process as well.

Yep. I can put together an issue template once we have to do this.


> 
> I do not think we should let CI make code changes, especially since this project has an unusual blend of requiring very deliberate, traceable changes from a maintainer base with limited bandwidth. Automatic fixes would mess with the existing squash-and-merge procedure as well as make it easier to have the (hopefully benign) unreviewed changes get merged accidentally.

I'm ok with the more conservative option for now; in my experience, I've only ever had it screw up once or twice, ever.

> To start I'd vote to have pre-commit added as a non-mandatory CI task so the results are visible to maintainers but not required while we tune the rule set. From there we can discuss if it makes sense to require a passing lint.

Sounds like a good way to phase it in. Should be straightforward to do.

> Since this would only impact the relatively few, relatively small repos owned by the Space ROS org and are worked on by a relatively small team I don't see a reason not to have a mass-fix commit and I think we'd want to preserve the traceability of when and why those changes were made rather than hiding it. The only complication would be if the checks turn up anything we don't want to change for whatever reason, but we'll need to handle that on a case-by-case basis either way.

Alrighty. Note that you can show "hidden" commits easily and still maintain full traceability as documented here: https://docs.github.com/en/repositories/working-with-files/using-files/viewing-a-file#bypassing-git-blame-ignore-revs-in-the-blame-view


---
Any tips on how you would like to do a vote here?

**Reply by EzraBrooks** [reply]
This all sounds great to me, feel free to start submitting PRs and we can discuss on a more granular level as appropriate :)

**Comment by Bckempa** [comment]
We're running out of time before the next release. I could implement this real fast but we would be short on time to PR any changes it recommends. Should we set it for the next release instead?

**Reply by Ryanf55** [reply]
Ah, thanks for the reminder. I can put up a PR tonight with the first round of hooks, which should be low impact and at least get the tooling running.

**Reply by Ryanf55** [reply]
First pass at https://github.com/space-ros/space-ros/pull/151
